### PR TITLE
UPSTREAM: 00000: wait for CRD discovery to be successful once before …

### PR DESCRIPTION
**Pull request by deads2k**
…reporting success

this is "simple" and inefficient.  I'm not completely sure that it will make it past the upstream review.  What do we think here.

### CODE REVIEW

**Comment by mfojtik** [comment]
Maybe it is simple and inefficient, but at least rule out the possibility of "server gives me 404 for my CR" and makes the cluster more reliable.

I would merge this now, see if the marketplace GC issue is fixed (or get better). If not, we can revert this. If it helps, we can then improve this in upstream and remove this carry.

/cc @sttts

File: vendor/k8s.io/kubernetes/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/apiserver.go [hunk 1]
```diff
@@ -199,6 +204,65 @@ func (c completedConfig) New(delegationTarget genericapiserver.DelegationTarget)
...
+					}
+
+					if !foundCRDVersion {
+						missingCRDs.Insert(crd.Spec.Names.Plural + "." + versionSpec.Name + "." + crd.Spec.Group)
+					}
+				}
+			}
+
+			if len(missingCRDs) > 0 {
+				glog.Warningf("waiting on CRD discovery for %v", missingCRDs.List())
```
Comment (sttts) [inline vendor/k8s.io/kubernetes/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/apiserver.go]: info

File: vendor/k8s.io/kubernetes/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/apiserver.go [hunk 2]
```diff
@@ -199,6 +204,65 @@ func (c completedConfig) New(delegationTarget genericapiserver.DelegationTarget)
...
+				return false, err
+			}
+			missingCRDs := sets.String{}
+			for _, crd := range crds {
+				// only check for those that should be in discovery
+				if !apiextensions.IsCRDConditionTrue(crd, apiextensions.Established) {
+					continue
+				}
+
+				for _, versionSpec := range crd.Spec.Versions {
```
Comment (sttts) [inline vendor/k8s.io/kubernetes/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/apiserver.go]: only check served versions

File: vendor/k8s.io/kubernetes/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/apiserver.go (hunk 1 above)
Comment (mfojtik) [reply vendor/k8s.io/kubernetes/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/apiserver.go]: I would want this to be clearly visible in the log (so we can confirm we indeed see missing CRD and without this hook we will return 404 to clients (and GC))

**Comment by mfojtik** [comment]
/hold

there is a bug

File: vendor/k8s.io/kubernetes/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/apiserver.go [hunk 3]
```diff
@@ -199,6 +204,69 @@ func (c completedConfig) New(delegationTarget genericapiserver.DelegationTarget)
...
+					if !versionSpec.Served {
+						continue
+					}
+
+					foundCRDVersion := false
+					for _, resourceList := range resourceLists {
```
Comment (sttts) [inline vendor/k8s.io/kubernetes/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/apiserver.go]: findAnyVersion:

File: vendor/k8s.io/kubernetes/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/apiserver.go [hunk 4]
```diff
@@ -199,6 +204,69 @@ func (c completedConfig) New(delegationTarget genericapiserver.DelegationTarget)
...
+							}
+							if resource.Name != crd.Spec.Names.Plural {
+								continue
+							}
+							if resource.Version != versionSpec.Name {
+								continue
+							}
+
+							foundCRDVersion = true
+							break
```
Comment (sttts) [inline vendor/k8s.io/kubernetes/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/apiserver.go]: break findAnyVersion

**Comment by mfojtik** [comment]
https://github.com/openshift/origin/pull/22149 replaces this.

**Pull request closed without merging**